janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/CustomAnalyzer.java

```diff
@@ -0,0 +1,121 @@
+
+/**
+ * @author Olivier binda (olivier.binda@wanadoo.fr)
+ */
+
+/**
+ * JanusGraph may open different index stores and needs an analyzer for each
+ * so an analyzer should be provided to build IndexWriters/IndexReaders that are used when an index is opened and documents are added
+ * BUT a janusGraph may add a new indexable field, with a custom tokenizing process through Parameters, at any moment, which means that it should be possible
+ * to map any field to the corresponding custom analyzer, though indexreaders are already in use.
+ * <p>
+ * A way to achieve that would be to use a delegating analyzer that delegates Tokenizing for a field to
+ * the corresponding Analyzer (whose class you get through a Mapping)
+ */
+public class CustomAnalyzer extends DelegatingAnalyzerWrapper {
+    private String store;
+    private KeyInformation.IndexRetriever informations;
+
+    private Map<String, Analyzer> analyzers = new HashMap<>();
+
+    public CustomAnalyzer(String store, KeyInformation.IndexRetriever informations, ReuseStrategy fallbackStrategy) {
+        super(fallbackStrategy);
+        this.store = store;
+        this.informations = informations;
+    }
+
+    @Override
+    protected Analyzer getWrappedAnalyzer(String fieldName) {
+        Mapping mapping = null;
+        String stringAnalyzerString = null;
+        String textAnalyzerString = null;
+
+
+        Parameter[] parameters = null;
+        Class<?> clazz = null;
+
+
+        KeyInformation keyInformation = informations.get(store, fieldName);
+        if (keyInformation != null) {
+            parameters = keyInformation.getParameters();
+            clazz = keyInformation.getDataType();
+        }
+
+        if (parameters != null) {
+            // if mapping isn't present in parameters, we use Mapping.DEFAULT
+            mapping = ParameterType.MAPPING.findParameter(parameters, Mapping.DEFAULT);
+            stringAnalyzerString = ParameterType.STRING_ANALYZER.findParameter(parameters, null);
+            textAnalyzerString = ParameterType.TEXT_ANALYZER.findParameter(parameters, null);
+        }
+      
+        String analyzerString = null;
+
+        if (clazz != null && String.class.isAssignableFrom(clazz) && mapping != null) {
+            // at the moment, we only try to support custom analyzers for string data.
+            // everything else falls through a StandardAnalyzer as was the case before
+            switch (mapping) {
+                case TEXTSTRING:
+                    throw new RuntimeException("TextString is an unsupported mapping for string data & custom analyzers");
+                case PREFIX_TREE:
+                    throw new RuntimeException("Prefix-tree is an unsupported mapping for string data & custom analyzers");
+                case STRING:
+                    if (stringAnalyzerString != null) analyzerString = stringAnalyzerString; else analyzerString = org.apache.lucene.analysis.core.KeywordAnalyzer.class.getName();
```

Comment:
`private static String KEYWORD_ANALYZER = org.apache.lucene.analysis.core.KeywordAnalyzer.class.getName();
analyzerString = stringAnalyzerString != null ? stringAnalyzerString : KEYWORD_ANALYZER;`

 will be more readable